pkg/driver/agent.go

```diff
@@ -79,11 +87,86 @@ func NewNode(d *CSIDriver) csi.NodeServer {
 		}
 	}()
 
+	if d.config.ListenAddress != "" {
+		exposeMetrics(d.config.ListenAddress, d.config.MetricsPath, d.config.DisableExporterMetrics)
+	}
+
 	return &node{
 		driver: d,
 	}
 }
 
+/*
+Function to register collectors to collect LVM related metrics and exporter metrics.
+
+If disableExporterMetrics is set to false, exporter will include metrics about itself i.e (process_*, go_*).
+*/
+func registerCollectors(disableExporterMetrics bool) (*prometheus.Registry, error) {
+	registry := prometheus.NewRegistry()
+
+	if !disableExporterMetrics {
+		processCollector := collectors.NewProcessCollector(collectors.ProcessCollectorOpts{})
+		err := registry.Register(processCollector)
+		if err != nil {
+			klog.Errorf("failed to register process collector for exporter metrics collection: %s", err.Error())
+			return nil, err
+		}
+		goProcessCollector := collectors.NewGoCollector()
+		err = registry.Register(goProcessCollector)
+		if err != nil {
+			klog.Errorf("failed to register go process collector for exporter metrics collection: %s", err.Error())
+			return nil, err
+		}
+	}
+	lvmCollector := collector.NewLvmCollector()
+
+	err := registry.Register(lvmCollector)
+	if err != nil {
+		klog.Errorf("failed to register LVM collector for LVM metrics collection: %s", err.Error())
+		return nil, err
+	}
+	return registry, nil
+}
+
+/*
+Function to start HTTP server to expose LVM metrics.
+
+Parameters:
+
+listenAddr: TCP network address where the prometheus metrics endpoint will listen.
+
+metricsPath: The HTTP path where prometheus metrics will be exposed.
+
+disableExporterMetrics: Exclude metrics about the exporter itself (process_*, go_*).
+*/
+func exposeMetrics(listenAddr string, metricsPath string, disableExporterMetrics bool) {
+
+	// Registry with all the collectors registered
+	registry, err := registerCollectors(disableExporterMetrics)
+	if err != nil {
+		klog.Fatalf("Failed to register collectors for LVM metrics collection: %s", err.Error())
+	}
+
+	http.Handle(metricsPath, promhttp.InstrumentMetricHandler(registry, promhttp.HandlerFor(registry, promhttp.HandlerOpts{
+		ErrorLog: log.NewErrorLogger(),
```

Comment:
Can we use klog here implementing relevant Logger interface?

Reply:
To implement the Logger interface with klog, We will have to implement the `Println(...)` method declared in the promhttp Logger interface with struct and then use it. So should I go with this approach or is there any other way possible?

Reply:
Have implemented Logger interface with klog.
Testing of logs:
- Without klog:
```
time="2021-06-18T09:28:01Z" level=error msg="error gathering metrics: error collecting metric Desc{fqName: \"lvm_vg_total_size_bytes\", help: \"LVM VG total size in bytes\", constLabels: {}, variableLabels: [name]}: encountered unknown type 5\n" source="log.go:181"
```

- With klog:
```
E0618 09:25:52.927741       1 agent.go:133] error gathering metrics:error collecting metric Desc{fqName: "lvm_vg_total_size_bytes", help: "LVM VG total size in bytes", constLabels: {}, variableLabels: [name]}: encountered unknown type 5
```